src/filter.ts

```diff
@@ -387,41 +387,47 @@ type NotUndefined = string | number | boolean | SomeObject;
  * Given a FilterQuery, returns an object that, when intersected with an Update,
  * marks those properties as required that are guaranteed to exist.
  */
-type RunQuery<Q extends string> = L1Combinations<Q, L1Parts<Q>>;
+type RunQuery<Q extends string> = L1Discriminator<Q, L1Parts<Q>>;
+
+// gets all L1 query snippets
+type L1Parts<Q extends string> = Q extends `${infer U}:${string}` ? U : Q;
+// gets all L2 query snippets for the given L1 part, or `never`
+type L2Parts<
+    Q extends string,
+    P extends string,
+> = Q extends `${P}:${infer U}:${string}` ? U
+    : Q extends `${P}:${infer U}` ? U
+    : never;
 
 // build up all combinations of all L1 fields
-type L1Combinations<Q extends string, L1 extends string> = Combine<
-    L1Fields<Q, L1>,
+type L1Discriminator<Q extends string, L1 extends string> = Combine<
+    L1Fragment<Q, L1>,
     L1
 >;
 // maps each L1 part of the filter query to an object
-type L1Fields<Q extends string, L1 extends string> = L1 extends unknown
-    ? Record<L1, L2Combinations<L2Parts<Q, L1>>>
+type L1Fragment<Q extends string, L1 extends string> = L1 extends unknown
+    ? Record<L1, L2Discriminator<L1, L2Parts<Q, L1>>>
     : never;
 
 // build up all combinations of all L2 fields
-type L2Combinations<L2 extends string> = [L2] extends [never] ? NotUndefined // short-circuit L1 queries (L2 is never)
-    : Combine<L2Fields<L2>, L2>;
+type L2Discriminator<L1 extends string, L2 extends string> = [L2] extends
+    [never] ? L2ShallowFragment<L1> // short-circuit L1 queries (L2 is never), only add twins
+    : Combine<L2Fragment<L1, L2>, L2>;
 // maps each L2 part of the filter query to an object and handles siblings
-type L2Fields<L2 extends string> = L2 extends unknown
-    ? Record<L2 | Twins<L2>, NotUndefined>
+type L2Fragment<L1 extends string, L2 extends string> = L2 extends unknown
+    ? Record<L2 | AddTwins<L1, L2>, NotUndefined>
     : never;
+// does the same as L1Fragment but without combining L2 properties
+type L2ShallowFragment<L1 extends string> = Record<
+    AddTwins<L1, never>,
+    NotUndefined
+>;
```

Comment:
Effectively only applied cross-level twins, could be implemented directly via `TwinsFromL1` but this is more readable and equivalent.